cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
-		BladeCLI bladeCLI = getBladeCLI();
+	private List<PathMatcher> _getPathMatchers(Path baseDirPath, String... patterns) {
+		FileSystem fileSystem = FileSystems.getDefault();
+
+		List<PathMatcher> pathMatchers = new ArrayList<>(patterns.length);
+
+		String patternPrefix = baseDirPath.toAbsolutePath() + File.separator;
+
+		if (File.separatorChar != '/') {
+			patternPrefix = patternPrefix.replace(File.separatorChar, '/');
+		}
 
-		BaseArgs baseArgs = bladeCLI.getArgs();
+		for (String pattern : patterns) {
+			if (pattern.startsWith("**/")) {
+				String absolutePattern = patternPrefix + pattern.substring(3);
 
-		File baseDir = new File(baseArgs.getBase());
+				_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+			}
 
-		WorkspaceProvider workspaceProvider = bladeCLI.getWorkspaceProvider(baseDir);
+			String absolutePattern = patternPrefix + pattern;
 
-		File workspaceDir = workspaceProvider.getWorkspaceDir(bladeCLI);
+			_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+		}
+
+		return pathMatchers;
+	}
 
-		Set<String> projectPaths = projectOutputFiles.keySet();
+	private Map<String, Path> _getProjectPaths(
+			final Path watchPath, List<String> projectPaths, List<String> ignorePaths)
+		throws Exception {
 
-		try (final GogoShellClient client = new GogoShellClient()) {
-			for (String projectPath : projectPaths) {
-				File projectDir = new File(workspaceDir, projectPath.replaceAll(":", File.separator));
+		final Map<String, Path> foundProjectPaths = new HashMap<>();
 
-				File watchOutputFile = new File(projectDir, "build/installedBundleId");
+		FileSystem fileSystem = FileSystems.getDefault();
 
-				if (watchOutputFile.exists()) {
-					try {
-						String installedBundleId = new String(Files.readAllBytes(watchOutputFile.toPath()));
+		List<PathMatcher> ignorePathMatchers = ignorePaths.stream(
+		).map(
+			ignorePath -> fileSystem.getPathMatcher("glob:" + ignorePath)
+		).collect(
+			Collectors.toList()
+		);
 
-						String response = client.send("uninstall " + installedBundleId);
+		Files.walkFileTree(
+			watchPath,
+			new SimpleFileVisitor<Path>() {
 
-						bladeCLI.out(response);
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					boolean shouldIgnorePath = ignorePathMatchers.stream(
+					).filter(
+						pathMatcher -> pathMatcher.matches(path)
+					).findFirst(
+					).isPresent();
+
+					if (shouldIgnorePath) {
+						return FileVisitResult.SKIP_SUBTREE;
 					}
-					catch (IOException ioe) {
+
+					try (Stream<Path> files = Files.list(path)) {
+						boolean projectPathFound = files.map(
+							p -> p.getFileName()
+						).filter(
+							p -> !ignorePathMatchers.stream(
+							).filter(
+								pathMatcher -> pathMatcher.matches(path.resolve(p))
+							).findFirst(
+							).isPresent()
+						).filter(
+							p -> projectPaths.stream(
```

Comment:
Can this be `).anyMatch(` instead? That eliminates the need for lines 209 - 210.

Reply:
Nice, that is more concise.